controllers/tests/controller_server/addons_server_test.py

```diff
@@ -41,34 +41,183 @@ def setUp(self):
+        self.mediator.get_replication.return_value = replication
+
+    def _prepare_request_params(self, replication_type, replication_name=REPLICATION_NAME):
+        if replication_type == REPLICATION_TYPE_MIRROR:
+            self.request.parameters = {PARAMETERS_SYSTEM_ID: SYSTEM_ID,
+                                       PARAMETERS_COPY_TYPE: COPY_TYPE}
+            replication_request = ReplicationRequest(OBJECT_INTERNAL_ID, OTHER_OBJECT_INTERNAL_ID, SYSTEM_ID, COPY_TYPE,
+                                                     REPLICATION_TYPE_MIRROR)
+        else:
+            self.request.parameters = {PARAMETERS_REPLICATION_POLICY: replication_name}
+            replication_request = ReplicationRequest(OBJECT_INTERNAL_ID, None, None, REPLICATION_COPY_TYPE_SYNC,
+                                                     REPLICATION_TYPE_EAR, replication_name)
+        return replication_request
+
+    def _test_enable_replication_succeeds(self, replication_type):
+        replication_request = self._prepare_request_params(replication_type)
+        self._prepare_replication_mocks()
 
         self.servicer.EnableVolumeReplication(self.request, self.context)
 
         self.assertEqual(grpc.StatusCode.OK, self.context.code)
-        replication_request = ReplicationRequest(OBJECT_INTERNAL_ID, OTHER_OBJECT_INTERNAL_ID, SYSTEM_ID, COPY_TYPE,
-                                                 REPLICATION_TYPE_MIRROR)
         self.mediator.get_replication.assert_called_once_with(replication_request)
         self.mediator.create_replication.assert_called_once_with(replication_request)
 
+    def _test_enable_replication_no_volume_fails(self):
+        self.mediator.get_object_by_id.return_value = None
+
+        self.servicer.EnableVolumeReplication(self.request, self.context)
+
+        self.assertEqual(grpc.StatusCode.NOT_FOUND, self.context.code)
+        self.mediator.get_replication.assert_not_called()
+        self.mediator.create_replication.assert_not_called()
+
+    def _test_enable_replication_idempotency(self, replication_type, replication_name=REPLICATION_NAME,
+                                             copy_type=COPY_TYPE, grpc_status=grpc.StatusCode.OK):
+        replication_request = self._prepare_request_params(replication_type, replication_name)
+        self._prepare_replication_mocks(replication_type=replication_type, copy_type=copy_type)
+
+        self.servicer.EnableVolumeReplication(self.request, self.context)
+
+        self.assertEqual(grpc_status, self.context.code)
+        self.mediator.get_replication.assert_called_once_with(replication_request)
+        self.mediator.create_replication.assert_not_called()
+
+    def _test_disable_replication_succeeds(self, replication_type):
+        replication_request = self._prepare_request_params(replication_type)
+        self._prepare_replication_mocks(replication_type=replication_type)
+
+        self.servicer.DisableVolumeReplication(self.request, self.context)
```

Comment:
you are in class `TestControllerServicerEnableVolumeReplication` not `Disable`
please split all the tests into the correct classes

Reply:
Added separate classes for Enable, disable, promote, demote and resync replication